src/persist/src/indexed/trace.rs

```diff
@@ -121,12 +177,98 @@ impl<K: Data, V: Data> BlobTrace<K, V> {
-            updates.push(blob.get_trace_batch(key)?);
+        for meta in self.batches.iter() {
+            updates.push(blob.get_trace_batch(&meta.key)?);
         }
         Ok(TraceSnapshot { ts_upper, updates })
     }
 
+    /// Merge two batches into one, forwarding all updates not beyond the current
+    /// `since` frontier to the `since` frontier.
+    fn merge<L: Blob>(
+        &mut self,
+        first: &BlobTraceBatchMeta,
+        second: &BlobTraceBatchMeta,
+        blob: &mut BlobCache<K, V, L>,
+    ) -> Result<BlobTraceBatchMeta, Error> {
+        if first.level != second.level {
+            return Err(format!(
+                "invalid merge of {:?} and {:?}: different compaction levels",
+                first, second
+            )
+            .into());
+        }
+
+        if first.desc.upper() != second.desc.lower() {
+            return Err(format!(
+                "invalid merge of {:?} and {:?}: non contiguous batches",
+                first, second,
+            )
+            .into());
+        }
+
+        let mut updates = vec![];
+
+        updates.extend(blob.get_trace_batch(&first.key)?.updates.iter().cloned());
+        updates.extend(blob.get_trace_batch(&second.key)?.updates.iter().cloned());
+
+        // TODO: use antichain more idiomatically.
+        let since_time = self.since[0];
+        for ((_, _), t, _) in updates.iter_mut() {
+            if *t < since_time {
+                *t = since_time;
+            }
+        }
+
+        // Consolidate
+        differential_dataflow::consolidation::consolidate_updates(&mut updates);
+
+        // Construct a new BlobTraceBatch
+        let desc = Description::new(
+            first.desc.lower().clone(),
+            second.desc.upper().clone(),
+            self.since.clone(),
+        );
+        let new_batch = BlobTraceBatch {
+            desc: desc.clone(),
+            updates,
+        };
+
+        let key = self.new_blob_key();
+        blob.set_trace_batch(key.clone(), new_batch)?;
+
+        // TODO actually clear the unwanted batches from the blob storage
+        Ok(BlobTraceBatchMeta {
+            key,
+            desc,
+            level: first.level + 1,
+        })
+    }
+
+    /// Take one step towards compacting the trace.
+    ///
+    /// Returns true if the trace was modified, false otherwise.
+    pub fn step<L: Blob>(&mut self, blob: &mut BlobCache<K, V, L>) -> Result<bool, Error> {
+        // TODO: should we remember our position in this list?
+        for i in 1..self.batches.len() {
+            if (self.batches[i - 1].level == self.batches[i].level)
+                && PartialOrder::less_equal(self.batches[i].desc.upper(), &self.since)
+            {
+                let first = self.batches[i - 1].clone();
+                let second = self.batches[i].clone();
+
+                let new_batch = self.merge(&first, &second, blob)?;
+
+                // TODO: more performant way to do this?
```

Comment:
Seems like you want a linked list? I've never used it but there seems to be one in the stdlib

You could also do it in two steps where you remove a thing and replace a thing which at least is only one memcopy of everything in the vec

Reply:
hmmm i think linked list would negatively affect the serialization to meta which is likely more perf sensitive than this step operation. I did the swap remove thing you asked so now the cost is one memcpy which seems like a fine place to stop in this pr and revisit if this ends up being a bottleneck

Reply:
Works for me!